discovery/manager.go

```diff
@@ -0,0 +1,300 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/prometheus/prometheus/config"
+
+	"github.com/prometheus/prometheus/discovery/azure"
+	"github.com/prometheus/prometheus/discovery/consul"
+	"github.com/prometheus/prometheus/discovery/dns"
+	"github.com/prometheus/prometheus/discovery/ec2"
+	"github.com/prometheus/prometheus/discovery/file"
+	"github.com/prometheus/prometheus/discovery/gce"
+	"github.com/prometheus/prometheus/discovery/kubernetes"
+	"github.com/prometheus/prometheus/discovery/marathon"
+	"github.com/prometheus/prometheus/discovery/openstack"
+	"github.com/prometheus/prometheus/discovery/triton"
+	"github.com/prometheus/prometheus/discovery/zookeeper"
+)
+
+// Discoverer provides information about target groups. It maintains a set
+// of sources from which TargetGroups can originate. Whenever a discovery provider
+// detects a potential change, it sends the TargetGroup through its channel.
+//
+// Discoverer does not know if an actual change happened.
+// It does guarantee that it sends the new TargetGroup whenever a change happens.
+//
+// Discoverers should initially send a full set of all discoverable TargetGroups.
+type Discoverer interface {
+	// Run hands a channel to the discovery provider(consul,dns etc) through which it can send
+	// updated target groups.
+	// Must returns if the context gets canceled. It should not close the update
+	// channel on returning.
+	Run(ctx context.Context, up chan<- []*config.TargetGroup)
+}
+
+// type pool struct {
+// 	cancel func()
+// 	tgroups []*config.TargetGroup
+// }
+
+// NewManager is the Discovery Manager constructor
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger:         logger,
+		actionCh:       make(chan func(context.Context)),
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		targets:        make(map[string]map[string][]*config.TargetGroup),
+		discoverCancel: []context.CancelFunc{},
+	}
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
+// Targets pool is kept in a map with a format map[targetSetName]map[providerName].
+type Manager struct {
```

Comment:
The docs on the internal maps should be moved above the fields. Otherwise they will show those implementation details in the regular Go doc page.s